[Ong Jing Chun] iP

jingchun19 opened the pull request:
# JacobMalon - Your Personal Task Manager 📋

A Java-based task management application that helps you stay organized. Built with **JavaFX** for a modern user interface and robust task tracking capabilities.

## Key Features 🚀
* Task Management
  * Create simple todos
  * Set deadlines with date/time
  * Schedule events with duration
* Smart Search
  * Find tasks by keywords
  * Filter by date
* Data Persistence
  * Auto-saves to `list.txt`
  * Maintains task status

## Getting Started
1. **Installation**
   1. Ensure Java 11 is installed
   2. Download the latest release
   3. Run `java -jar jacobmalon.jar`
2. **Basic Commands**
   1. `todo read book` - Create a todo
   2. `deadline return book /by 2/12/2023 1800` - Set a deadline
   3. `event team meeting /from 2/12/2023 1400 /to 2/12/2023 1600` - Schedule an event

### Example Usage

## Project Status
- [x] Basic Task Operations
- [x] GUI Implementation
- [x] File Storage
- [x] Date/Time Validation
- [ ] Cloud Sync
- [ ] Mobile App Integration

> "The best way to get something done is to begin." - *Unknown*

## Technologies Used
* *Java 11*
* **JavaFX**
* ~~Command Line Interface~~ (replaced with GUI)
* __File-based Storage__

For more information about the project, visit our [GitHub Repository](https://github.com/yourusername/jacobmalon).

## Acknowledgments
Special thanks to:
* CS2103T Teaching Team
* JavaFX Community
* All contributors

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jingchun19 committed:
Rename to JacobMalon and created greet

jingchun19 committed:
level 0 changes

jingchun19 committed:
Echo created

jingchun19 committed:
Level 2 changes Add, List

jingchun19 committed:
Level 3

jingchun19 committed:
Level-4

jingchun19 committed:
Test file

jingchun19 committed:
Handle Errors, mark and todo

jingchun19 committed:
delete function

jingchun19 committed:
Test changes

jingchun19 committed:
main edit

jingchun19 committed:
Added text file

jingchun19 committed:
save function

jingchun19 committed:
Added date management, works for YYYY-MM-DD

jingchun19 committed:
Date format

aquaimpact commented on src/main/java/JacobMalon.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Scanner;
+import taskscommand.*;
+
+public class JacobMalon {
```
Great job on the implementation! 🎉 To improve readability and maintainability, I suggest adding Javadoc comments for both your classes and methods. This will help clarify their purpose, expected inputs, return values, and any potential exceptions they might throw. Providing detailed documentation will also make it easier for others (and your future self!) to understand and use your code effectively. Keep up the good work! 👍

aquaimpact commented on src/main/java/JacobMalon.java:
```diff
@@ -0,0 +1,64 @@
+                        break;
+                    } else if (command.equals("list")) {
+                        taskManager.listTasks();
+                    } else if (command.startsWith("mark ")) {
```
Interesting use of the startsWith() command! I didn't think of that!

aquaimpact commented on src/main/java/JacobMalon.java:
```diff
@@ -0,0 +1,64 @@
+                    } else if (command.startsWith("deadline ")) {
+                        String[] parts = command.substring(9).split(" /by ", 2);
+                        if (parts.length < 2) {
+                            throw new IllegalArgumentException("deadline command must be followed by description and date (e.g., deadline return book /by 2019-10-15)");
```
Good use of `IllegalArgumentException()`! However, you might consider defining a custom task-related exception instead. This would make error handling more specific and prevent potential confusion in the future if other parts of the codebase also throw `IllegalArgumentException()` for different reasons. A custom exception would improve clarity, maintainability, and debugging.

aquaimpact commented on src/main/java/taskscommand/TaskManager.java:
```diff
@@ -0,0 +1,135 @@
+        } catch (IOException e) {
+            System.out.println("Error saving to file: " + e.getMessage());
+        }
+    }
```
You might consider abstracting this chunk of code responsible for saving data into a separate class, such as `SaveFile.java`. This would improve code readability, maintainability, and separation of concerns, making it easier to understand and debug in the future.

aquaimpact commented on src/main/java/taskscommand/TaskManager.java:
```diff
@@ -0,0 +1,135 @@
+    public void listTasksByDate(String date) {
+        try {
+            LocalDate queryDate = LocalDate.parse(date, INPUT_FORMATTER);
+            List<Task> tasksOnDate = tasks.stream()
```
Interesting use of stream()! I did not think of that!

aquaimpact commented on src/main/java/taskscommand/Deadline.java:
It would be beneficial to implement JUnit tests for this and other files to verify that each function runs correctly when given different inputs. This will help ensure the stability of your project, especially as you introduce more modifications in the future.

aquaimpact reviewed (commented):
Overall LGTM. Just need to fix several parts of your code and you are all good! Do also ensure that you follow the checkstyle closely!

jingchun19 committed:
Update .gitignore

jingchun19 committed:
Create .gitattributes

jingchun19 committed:
JavaDoc

jingchun19 committed:
Merge branch 'master' of https://github.com/jingchun19/ip

jingchun19 committed:
Gradle files

jingchun19 committed:
Issues with Junit testing

jingchun19 committed:
jar file

jingchun19 committed:
find function

Fieash reviewed (commented):
- could be better titled as add date support
- can it support time or other date formats? e.g. 12.12.2025 1800

jingchun19 committed:
Add caches/ to .gitignore

jingchun19 committed:
gradle

jingchun19 committed:
Save changes for daemon log file

jingchun19 committed:
Add daemon/ to .gitignore

jingchun19 committed:
Remove daemon files from tracking

jingchun19 committed:
Update Task.java

jingchun19 committed:
Ignore caches, daemon, native, and wrapper directories

jingchun19 committed:
branch level 10 changes

jingchun19 committed:
Commit changes on branch-level-10

jingchun19 committed:
gitignore

jingchun19 committed:
Merge branch 'branch-level-10'

jingchun19 committed:
Coding Standards

jingchun19 committed:
Level-9 changes

jingchun19 committed:
JavaDoc

jingchun19 committed:
Merge branch-A-CodingStandard with --no-ff

jingchun19 committed:
Merge branch-A-JavaDoc with --no-ff

jingchun19 committed:
Merge branch 'branch-Level-9'

jingchun19 committed:
added native caches and daemon file to gitignore

jingchun19 committed:
Remove daemon log file from tracking

jingchun19 committed:
Remove daemon folder from tracking

jingchun19 committed:
Ignore caches and wrapper directories

jingchun19 committed:
Remove runtest.sh

jingchun19 committed:
Test for Deadline and Task

jingchun19 committed:
GFMD added

jingchun19 committed:
Implement assertions

jingchun19 committed:
Improve code quality

jingchun19 committed:
Merge pull request #3 from jingchun19/branch-A-CodeQuality

Improve code quality

jingchun19 committed:
Merge pull request #2 from jingchun19/branch-A-Assertions

Implement assertions

jingchun19 committed:
Add assertions to improve code robustness

Task-related classes lack defensive programming measures to catch
programming errors early in development.

Let's add assertions to:
* Validate task descriptions are non-null and non-empty
* Ensure deadline dates are properly parsed
* Guard against invalid task indices
* Verify task list integrity after operations
* Protect against null parameters in public methods

These assertions help catch programming errors during development
and make the code's assumptions explicit. They complement existing
exception handling which deals with runtime errors.

Part of code quality improvements for ip project.

jingchun19 committed:
Improve code quality of task management classes

Task management classes have several code quality issues including deep
nesting, long methods, and unclear variable names.

Let's improve the code quality by:
* Extract helper methods to reduce method length and nesting
* Add meaningful constants for magic strings
* Group related operations together
* Improve method and variable naming
* Separate concerns between task and file operations
* Add comprehensive JavaDoc comments
* Apply SLAP (Single Level of Abstraction Principle)

These changes make the code more maintainable and readable while
preserving the existing functionality. Methods now follow a clear
logical structure and operate at consistent abstraction levels.

The improvements align with the coding quality guidelines:
- Methods are kept under 30 LOC
- Nesting is limited to 2-3 levels
- Each method has a single responsibility
- Code is self-documenting where possible

Part of ongoing code quality improvements for ip project.

jingchun19 committed:
Pull from origin

jingchun19 committed:
Add duplicate task detection

Tasks can currently be added multiple times without any warning,
which may lead to confusion and cluttered task lists.

Let's add duplicate detection that:
* Identifies duplicate tasks based on their content
* Warns users when they try to add a duplicate task
* Handles different task types appropriately:
  - Todo: checks description equality
  - Deadline: checks description and deadline timing
  - Event: checks description and event timing

The implementation allows duplicates to be added after warning
the user, providing flexibility while preventing accidental
duplicates.

Part of C-DetectDuplicates extension implementation.

jingchun19 committed:
Update User Guide in docs/README.md

jingchun19 committed:
Added UI.png

jingchun19 committed:
Corrected save and load file

jingchun19 committed:
Added closing by saying bye

jingchun19 committed:
Update test files for task storage and formatting

- Add tests for file I/O operations in TaskManager
- Test file creation when storage file doesn't exist
- Verify correct date/time formatting for tasks
- Test save/load functionality with different task types
- Add test cases for input format consistency
- Ensure proper handling of empty storage file

What's changed:
- Added new test cases for file operations
- Updated existing tests to match new date format
- Added edge case tests for file handling
- Improved test coverage for task formatting

jingchun19 committed:
Standardize date/time handling across application

- Implement consistent date/time format validation
- Use shared INPUT_FORMATTER (d/M/yyyy HHmm) across all classes
- Add proper date parsing and validation in JacobMalon
- Improve error messages with format examples
- Add validation for event start/end time order
- Update file I/O to maintain date format consistency

What's changed:
- JacobMalon now validates dates before creating tasks
- Consistent date format validation across Deadline and Event
- Better error messages with examples
- Proper handling of invalid date formats

jingchun19 committed:
standardize: Unify date/time handling across app

Date/time formats are inconsistent across different parts of the app.
This causes confusion for users and makes the code harder to maintain.

Let's standardize the date/time handling by:
* Using shared INPUT_FORMATTER (d/M/yyyy HHmm) across all classes
* Adding proper validation before task creation
* Improving error messages with format examples
* Validating event start/end time order

This standardization:
* Makes the user experience more consistent
* Reduces code duplication
* Improves error handling
* Makes the codebase more maintainable

The validation is done at input level rather than entity level to fail
fast and provide better user feedback.

jingchun19 committed:
feat: Add persistent storage for tasks

Tasks are lost when the application is closed. Users need their tasks
to persist between sessions to make the app useful.

Let's implement file-based storage that:
* Creates list.txt if it doesn't exist
* Saves tasks in user input format
* Loads tasks on application startup
* Handles different task types (Todo/Deadline/Event)

This implementation:
* Preserves exact user input format for easier editing
* Maintains task status across sessions
* Provides graceful handling of missing storage file
* Ensures backward compatibility

The storage format matches user input commands to maintain consistency
between what users type and what they see in the file.

jingchun19 committed:
docs: Update error messages with clear examples

Error messages lack specific examples, making it difficult for users
to understand the correct input format. This leads to user frustration
and increased support queries.

Let's improve error messages by:
* Adding format examples (e.g., 2/12/2023 1800)
* Clarifying command syntax with real-world examples
* Standardizing message format across all commands
* Including complete command examples

Examples now show:
* Todo: todo read book
* Deadline: deadline return book /by 2/12/2023 1800
* Event: event team meeting /from 2/12/2023 1400 /to 2/12/2023 1600

Better error messages reduce user confusion and make the application
more user-friendly for first-time users.

jingchun19 committed:
test: Add validation tests for date/time formats

Test coverage for date/time validation is incomplete. Invalid formats
could slip through, causing runtime errors and poor user experience.

Let's add comprehensive test cases for:
* Valid date formats (d/M/yyyy HHmm)
* Invalid date patterns
* Edge cases (single digit dates/months)
* Event time order validation
* Boundary conditions (year transitions)

These tests ensure:
* Consistent date/time handling
* Proper error messages for invalid inputs
* Robust validation across all task types
* Prevention of invalid event time ranges

Added both unit and integration tests to verify the entire validation
pipeline from user input to task creation.

jingchun19 committed:
refactor: Extract date formatter to utility class

Date formatters are duplicated across Task classes, violating DRY
principle. Changes to date format require updates in multiple places.

Let's create a DateTimeUtil class that:
* Centralizes all date/time formatting logic
* Provides shared INPUT_FORMATTER and OUTPUT_FORMATTER
* Adds validation methods for date strings
* Handles parsing exceptions consistently

Benefits:
* Single source of truth for date formats
* Easier maintenance and format changes
* Consistent error handling
* Reduced code duplication

The utility class is package-private to maintain encapsulation while
allowing reuse within the tasks package.

jingchun19 committed:
fix: Rename UI.png to Ui.png for standard compliance

Image naming doesn't follow project conventions where only the first
letter should be capitalized. This inconsistency affects documentation
and README rendering.

Let's:
* Rename UI.png to Ui.png
* Update all references to the image
* Maintain consistent naming convention
* Ensure proper image loading in documentation

This change aligns with:
* Project naming conventions
* Java class naming standards
* Documentation best practices

The new filename matches other UI-related components in the project,
improving overall code consistency.
